lxd/storage/drivers/drivers_lvm_volumes.go

```diff
@@ -0,0 +1,732 @@
+		}
+	} else {
+		err = d.resizeLogicalVolume(volDevPath, newSizeBytes)
+		if err != nil {
+			return err
+
+		}
+	}
+
+	return nil
+}
+
+// CreateVolumeSnapshot creates a snapshot of a volume.
+func (d *lvm) CreateVolumeSnapshot(snapVol Volume, op *operations.Operation) error {
+	parentName, _, _ := shared.InstanceGetParentAndSnapshotName(snapVol.name)
+	parentVol := NewVolume(d, d.name, snapVol.volType, snapVol.contentType, parentName, snapVol.config, snapVol.poolConfig)
+	snapPath := snapVol.MountPath()
+
+	// Create the parent directory.
+	err := createParentSnapshotDirIfMissing(d.name, snapVol.volType, parentName)
+	if err != nil {
+		return err
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Create snapshot directory.
+	err = snapVol.EnsureMountPath()
+	if err != nil {
+		return err
+	}
+	revert.Add(func() { os.RemoveAll(snapPath) })
+
+	_, err = d.createLogicalVolumeSnapshot(d.config["lvm.vg_name"], parentVol, snapVol, true, d.usesThinpool())
+	if err != nil {
+		return fmt.Errorf("Error creating LVM logical volume snapshot: %v", err)
+	}
+
+	revert.Success()
+	return nil
+}
+
+// DeleteVolumeSnapshot removes a snapshot from the storage device. The volName and snapshotName
+// must be bare names and should not be in the format "volume/snapshot".
+func (d *lvm) DeleteVolumeSnapshot(snapVol Volume, op *operations.Operation) error {
+	// Remove the snapshot from the storage device.
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if lvExists {
+		_, err = d.UnmountVolume(snapVol, op)
+		if err != nil {
+			return fmt.Errorf("Error unmounting LVM logical volume: %v", err)
+		}
+
+		err = d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name))
+		if err != nil {
+			return fmt.Errorf("Error removing LVM logical volume: %v", err)
+		}
+	}
+
+	// Remove the snapshot mount path from the storage device.
+	snapPath := snapVol.MountPath()
+	err = os.RemoveAll(snapPath)
+	if err != nil {
+		return errors.Wrapf(err, "Error removing LVM snapshot mount path '%s'", snapPath)
+	}
+
+	// Remove the parent snapshot directory if this is the last snapshot being removed.
+	parentName, _, _ := shared.InstanceGetParentAndSnapshotName(snapVol.name)
+	err = deleteParentSnapshotDirIfEmpty(d.name, snapVol.volType, parentName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RenameVolumeSnapshot renames a volume snapshot.
+func (d *lvm) RenameVolumeSnapshot(snapVol Volume, newSnapshotName string, op *operations.Operation) error {
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if !lvExists {
+		return fmt.Errorf("LVM logical volume doesn't exist")
+	}
```

Comment:
Why is this needed?

Reply:
@tomponline got this one to address still

Reply:
Yep, working my way down, removed this now.